libvast/builtins/operators/shell.cpp

```diff
@@ -0,0 +1,209 @@
+        if (not child.running(ec)) {
+          if (ec)
+            VAST_DEBUG(ec);
+          else
+            VAST_ERROR(ec);
+          co_yield {};
+          break;
+        }
+        // Pass operator input to the child's stdin.
+        const auto* chunk_data = reinterpret_cast<const char*>(chunk->data());
+        auto chunk_size = detail::narrow_cast<std::streamsize>(chunk->size());
+        VAST_DEBUG("writing {} bytes to child's stdin", chunk_size);
+        if (not child_stdin.write(chunk_data, chunk_size)) {
+          ctrl.abort(
+            caf::make_error(ec::unspecified,
+                            fmt::format("failed to write into child's stdin")));
+          co_yield {};
+          break;
+        }
+        // Try yielding so far accumulated child output.
+        std::unique_lock lock{chunks_mutex, std::try_to_lock};
+        if (lock.owns_lock()) {
+          size_t i = 0;
+          auto total = chunks.size();
+          while (not chunks.empty()) {
+            auto chk = chunks.front();
+            chunks.pop();
+            VAST_DEBUG("yielding chunk {}/{} with {} bytes", ++i, total,
+                       chk->size());
+            co_yield chk;
+          }
+        } else {
+          co_yield {};
+        }
+      }
+    }
+    // Yield all accumulated child output.
+    std::lock_guard lock{chunks_mutex};
+    size_t i = 0;
+    auto total = chunks.size();
+    while (not chunks.empty()) {
+      auto chk = chunks.front();
+      VAST_DEBUG("yielding chunk {}/{} with {} bytes", ++i, total, chk->size());
+      co_yield chk;
+      chunks.pop();
+    }
+  }
+
+  auto to_string() const -> std::string override {
+    return fmt::format("shell \"{}\"", command_);
+  }
+
+  auto location() const -> operator_location override {
+    // The user expectation is that shell executes relative to the currently
+    // executing process.
+    return operator_location::local;
+  }
+
+  auto detached() const -> bool override {
+    // We may execute blocking syscalls.
+    return true;
+  }
+
+private:
+  std::string command_;
+};
+
+class plugin final : public virtual operator_plugin {
+public:
+  // plugin API
+  auto initialize([[maybe_unused]] const record& plugin_config,
+                  [[maybe_unused]] const record& global_config)
+    -> caf::error override {
+    return {};
+  }
+
+  [[nodiscard]] auto name() const -> std::string override {
+    return "shell";
+  };
+
+  auto make_operator(std::string_view pipeline) const
+    -> std::pair<std::string_view, caf::expected<operator_ptr>> override {
+    using parsers::optional_ws_or_comment, parsers::required_ws_or_comment,
+      parsers::end_of_pipeline_operator, parsers::operator_arg;
+    const auto* f = pipeline.begin();
+    const auto* const l = pipeline.end();
+    const auto p = -(required_ws_or_comment >> operator_arg)
+                   >> optional_ws_or_comment >> end_of_pipeline_operator;
+    std::string command;
+    if (not p(f, l, command)) {
```

Comment:
This doesn't seem to handle arguments for the command, which is a severe limitation:
```fish
> journalctl -e -o json | head -n 1 | build/clang/debug/bin/vast exec 'load stdin | shell jq -n 4 | save stdout'
!! invalid_argument: failed to parse pipeline: !! syntax_error: invalid query: !! unspecified: failed to parse pipeline ' shell jq -n 4 | save stdout': !! syntax_error: failed to parse shell operator: ' jq -n 4 | save stdout'

> journalctl -e -o json | head -n 1 | build/clang/debug/bin/vast exec 'load stdin | shell "jq -n 4" | save stdout'                                                                                               vast-dev-env 
4
```

Reply:
Allowing any argument here is dangerous from a language perspective. I would rather forbid it right now and then consider removing the required quotes in the future, but any such thing should first go through @jachris.

Reply:
The second example doesn't work either. `jq -n 4` should set the indentation level to 4, not print a single "4".

Reply:
`jq -n 4` just prints `4` for me on the command line as well. If I run your example with `jq --indent 4` it works as you'd expect.

From `man jq`:

> `--indent n`:
> Use the given number of spaces (no more than 8) for indentation.

Reply:
It seems my brain turned the `n` into a shortopt while glancing at that man page.